lib/charms/opensearch/v0/helper_conf_setter.py

```diff
@@ -0,0 +1,303 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Utilities for editing yaml config files at any depth level and maintaining comments."""
+import re
+import sys
+import uuid
+from collections.abc import Mapping
+from enum import Enum
+from io import StringIO
+from os.path import exists
+from typing import Dict, List
+
+from ruamel.yaml import YAML, CommentedSeq
+from ruamel.yaml.comments import CommentedSet
+
+# The unique Charmhub library identifier, never change it
+LIBID = "69a1559c3e4c40cebd8ff0b255cf13db"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+class OutputType(Enum):
+    """Enum representing the output type of a write operation."""
+
+    file = "file"
+    obj = "obj"
+    console = "console"
+    all = "all"
+
+    def __str__(self):
+        """String representation of enum value."""
+        return self.value
+
+
+class YamlConfigSetter:
+    """Utility class for updating YAML config, supporting diverse object types and nestedness.
+
+    conf_setter = ConfigSetter()
+    put("file.yml", "a.b", "new_name")
+    put("file.yml", "a.b/c.obj/key3/key1.a/obj", {"a": "new_name_1", "b": ["hello", "world"]})
+    put("file.yml", "a.b/c.arr.simple/[0]", "hello")
+    put("file.yml", "a.b/c.arr.complex/[name:name1]", {"a": "new_name_1", "b": ["hello", "world"]})
+    put("file.yml", "a.b/c.arr.complex/[name:name5]/val/key", "new_val25")
+    put("file.yml", "a.b/c.arr.complex/[0]/val", "complex2_updated")
+    put("file.yml", "a.b/c.arr.complex/[0]/new_val/new_sub_key", "updated")
+    """
+
+    def __init__(self, base_path: str = None):
+        """base_path: if set, where to look for files relatively on "load/put/delete" methods."""
+        self.yaml = YAML()
+        self.base_path = self.__clean_base_path(base_path)
+
+    def load(self, config_file: str) -> Dict[str, any]:
+        """Load the content of a YAML file."""
+        path = f"{self.base_path}{config_file}"
+
+        if not exists(path):
+            raise FileNotFoundError(f"{path} not found.")
+
+        with open(path, "r") as f:
+            lines = f.read().splitlines()
+
+            random_id = uuid.uuid4().hex
+            lines.append(f"{random_id}: {random_id}")
```

Comment:
Just out of curiosity: why is this random id needed (and removed later)?

Reply:
This handles the case when a YAML document is only filled with comments. In that specific case, Ruamel (the parser) struggles to understand that this is indeed a YAML document, so it does not load the comments in its internal data structure, so when new data gets written in the file, all the comments are gone. 
This random id is here to give the parser a hint that it is a YAML document, so it preserves all the commented code, then once loaded - this id no longer serves a purpose, so it gets deleted

Reply:
Nice, thanks for the explanation!